### PR TITLE
[Packaging] RPM generation

### PR DESCRIPTION
**Pull request by artanokhov**
### Details:
 - Extend existing Cmake install/Cpack logic with RPM generation

### Tickets:
 - 90199

### CODE REVIEW

File: cmake/developer_package/packaging/rpm.cmake [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+#
+# Common Debian specific settings
+#
+
+macro(ov_rpm_specific_settings)
+    # multiple packages are generated
+    set(CPACK_RPM_COMPONENT_INSTALL ON)
+    # automatically find dependencies for binaries
+    set(CPACK_RPM_PACKAGE_SHLIBDEPS ON)
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm.cmake:77]: there is no such var

File: cmake/developer_package/packaging/rpm.cmake [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    # enable dependencies between components
+    set(CPACK_RPM_ENABLE_COMPONENT_DEPENDS ON)
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm.cmake:79]: no such var

File: cmake/developer_package/packaging/rpm.cmake [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    # control file permissions
+    set(CPACK_RPM_PACKAGE_CONTROL_STRICT_PERMISSION OFF)
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm.cmake:81]: no such var

File: cmake/developer_package/packaging/rpm.cmake [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    # homepage
+    set(CPACK_RPM_PACKAGE_HOMEPAGE "https://docs.openvino.ai/")
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm.cmake:83]: no such var

File: cmake/developer_package/packaging/rpm.cmake [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    # WA: dpkg-shlibdeps requires folder with libraries
+    # proper way is to use -l (path to libs) and -L (path to shlibs) for other already installed components
+    # but it requires CMake source code changes
+    # with current WA automatic deps detection via dpkg-shlibdeps for "our libraries"
+    # is ignored; but dependencies between our components are here because of
+    # CPACK_COMPONENT_<UCOMP>_DEPENDS variables
+    # More proper WA is try to enable INSTALL_RPATH
+
+    if(DEFINED CMAKE_LIBRARY_OUTPUT_DIRECTORY)
+        set(CPACK_RPM_PACKAGE_SHLIBDEPS_PRIVATE_DIRS "${CMAKE_LIBRARY_OUTPUT_DIRECTORY}")
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm.cmake:98]: no such var
need to understand how dependencies are searched in rpm case

File: cmake/developer_package/packaging/rpm.cmake [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    else()
+        message(FATAL_ERROR "CMAKE_LIBRARY_OUTPUT_DIRECTORY is empty")
+    endif()
+
+    # automatic dependencies discovering between openvino and user packages
+    set(CPACK_RPM_PACKAGE_GENERATE_SHLIBS ON)
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm.cmake:104]: no such var

File: cmake/developer_package/packaging/rpm.cmake [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    # OpenVINO does not have backward and forward compatibility
+    set(CPACK_RPM_PACKAGE_GENERATE_SHLIBS_POLICY "=")
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm.cmake:106]: no such var

File: cmake/developer_package/packaging/rpm.cmake [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+    # naming convention for rpm package files
+    set(CPACK_RPM_FILE_NAME "DEB-DEFAULT")
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm.cmake:108]: deb?

File: cmake/developer_package/packaging/rpm_post_build.cmake [hunk 9]
```diff
@@ -0,0 +1,41 @@
+# Copyright (C) 2018-2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+find_program(rpmlint_PROGRAM NAMES rpmlint DOC "Path to rpmlint tool")
+if(NOT rpmlint_PROGRAM)
+    message(WARNING "Failed to find 'rpmlint' tool, use 'sudo apt-get install rpmlint' to install it")
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm_post_build.cmake:7]: apt-get ?

File: cmake/packaging/rpm.cmake [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+        set(CPACK_RPM_HETERO_PACKAGE_CONTROL_EXTRA "${def_postinst};${def_postrm}")
+        _ov_add_plugin(hetero ON)
+    endif()
+
+    # auto batch
+    if(ENABLE_AUTO_BATCH)
+        set(CPACK_COMPONENT_BATCH_DESCRIPTION "OpenVINO Automatic Batching plugin")
+        set(CPACK_COMPONENT_BATCH_DEPENDS "core")
+        set(CPACK_RPM_BATCH_PACKAGE_NAME "libopenvino-auto-batch-${cpack_name_ver}")
+        set(CPACK_RPM_BATCH_PACKAGE_CONTROL_EXTRA "${def_postinst};${def_postrm}")
```
Comment (ilya-lavrenov) [inline cmake/packaging/rpm.cmake:130]: there is no such var in PRM generator

**Review by ilya-lavrenov — COMMENTED** [review]
In general, need read https://cmake.org/cmake/help/latest/cpack_gen/rpm.html and adjust according to RPM generator. There is no 1:1 mapping for each variable

File: cmake/developer_package/packaging/rpm.cmake (hunk 5 above)
Comment (artanokhov) [reply cmake/developer_package/packaging/rpm.cmake:96]: There is CPACK_RPM_PACKAGE_AUTOREQPROV for automatic listing of shared libraries and automatic shared libraries dependency detection at the same time.
https://cmake.org/cmake/help/v3.8/module/CPackRPM.html?highlight=cpack_rpm_package#variable:CPACK_RPM_PACKAGE_AUTOREQPROV

File: cmake/developer_package/packaging/rpm.cmake [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    # enable this if someday we change the version scheme
+    # set(CPACK_RPM_PACKAGE_EPOCH "2")
+endmacro()
+
+ov_rpm_specific_settings()
+
+# needed to override cmake auto generated files
+set(def_postinst "${OpenVINO_BINARY_DIR}/_CPack_Packages/postinst")
+set(def_postrm "${OpenVINO_BINARY_DIR}/_CPack_Packages/postrm")
+set(def_triggers "${OpenVINO_BINARY_DIR}/_CPack_Packages/triggers")
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm.cmake:100]: does rpm have the same format of postinst scripts? What if we remove them?

File: cmake/developer_package/packaging/rpm.cmake [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+            COMPONENT ${comp})
+
+    # create changelog.gz
+
+    find_host_program(gzip_PROGRAM NAMES gzip DOC "Path to gzip tool")
+    if(NOT gzip_PROGRAM)
+        message(FATAL_ERROR "Failed to find gzip tool")
+    endif()
+
+    set(changelog_src "${OpenVINO_SOURCE_DIR}/cmake/developer_package/packaging/changelog")
```
Comment (ilya-lavrenov) [inline cmake/developer_package/packaging/rpm.cmake:139]: does RPM has the same format of change log?

File: cmake/packaging/rpm.cmake [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+    set(CPACK_RPM_CORE_PACKAGE_NAME "libopenvino-${cpack_name_ver}")
+    # we need triggers to run ldconfig for openvino
+    set(CPACK_RPM_CORE_PACKAGE_CONTROL_EXTRA "${def_postinst};${def_postrm};${def_triggers}")
+    # use lintian to check packages in post-build step
+    set(CPACK_POST_BUILD_SCRIPTS "${IEDevScripts_DIR}/packaging/rpm_post_build.cmake")
+
+    # We currently don't have versioning for openvino core library
+    ov_rpm_add_rpmlint_suppression(core
+        "shlib-without-versioned-soname"
+        "package-name-doesnt-match-sonames")
```
Comment (ilya-lavrenov) [inline cmake/packaging/rpm.cmake:100]: does it have the same error messages as lintian? maybe we need to remove it? or suppress different tags

**Review by ilya-lavrenov — APPROVED** [review]
Good as initial version. All improvements can be done later

Please, replicate condition for DEB on RPM as well https://github.com/openvinotoolkit/openvino/search?q=CPACK_GENERATOR&type=code

**Pull request merged**